Sources/SwiftFormatPrettyPrint/PrettyPrint.swift

```diff
@@ -43,66 +31,129 @@ public class PrettyPrinter {
 
   /// This is set to true when a break creates a new line. Since consecutive break tokens may not
   /// all create new lines, only syntax tokens set this to false.
-  private var lastBreakConsecutive = false
+  private var lastBreakConsecutive = true
 
   /// Did the previous token create a new line? This is used to determine if a group needs to
   /// consistently break.
   private var lastBreak = false
 
-  /// The offset value of the last break token.
-  private var lastBreakOffset = 0
-
-  /// The total number of spaces we need to indent from the last break.
-  private var lastBreakValue = 0
-
-  /// Keep track of the indentation level of the current group as the number of blank spaces.
-  private var indentStack = [0]
-
-  /// Keep track of the offset value passed to a group. This is used to adjust the offset value of
-  /// any trailing break statements in the group.
-  private var relativeIndentStack = [0]
+  /// Keep track of the indentation level of the current group as a sequence of space or tab values.
+  private var indentationStack: [Indent] = []
 
   /// Keep track of whether we are forcing breaks within a group (for consistent breaking).
   private var forceBreakStack = [false]
 
-  /// If true, the pretty printer will output control characters that indicate where groups and
-  /// breaks occur in the formatted output.
-  private var isDebugMode: Bool
-
   /// If true, the token stream is printed to the console for debugging purposes.
   private var printTokenStream: Bool
 
-  /// The current index (0..<6) of the color to be used for the next color group.
-  private var currentDebugGroupMarkerColor = 0
+  /// Keeps track of the line numbers of the open (and unclosed) breaks seen so far.
+  private var openDelimiterBreakStack: [Int] = []
+
+  /// Keeps track of the current line number being printed.
+  private var lineNumber: Int = 0
 
-  /// We cycle through ANSI colors 31...36 for group brackets.
-  private let groupMarkerColorCount = 6
+  /// Indicates whether or not the current line being printed is a continuation line.
+  private var currentLineIsContinuation = false
 
-  /// The ANSI color code string for the current group color (used in debug mode).
-  private var currentGroupColorString: String {
-    return Ansi.color(currentDebugGroupMarkerColor + 1)
+  /// Keeps track of the most recent number of consecutive newlines that have been printed.
+  ///
+  /// This value is reset to zero whenever non-newline content is printed.
+  private var consecutiveNewlineCount = 0
```

Comment:
Hmm, this might be redundant now. I can try replacing it and seeing if everything still works as expected.

Reply:
Good call, I was able to remove that flag entirely. It turns out I don't even need to check `consecutiveNewlineCount` where we used to check the old flag, because `writeNewlines` handles all the cases naturally (if it computes zero, it just does nothing).